nisystemlink/clients/testmonitor/_test_monitor_client.py

```diff
@@ -23,6 +37,159 @@ def api_info(self) -> models.ApiInfo:
+            failures.
+
+        Raises:
+            ApiException: if unable to communicate with the ``/nitestmonitor`` service of provided invalid
+                arguments.
+        """
+        ...
+
+    @get(
+        "products",
+        args=[Query("continuationToken"), Query("take"), Query("returnCount")],
+    )
+    def get_products(
+        self,
+        continuation_token: Optional[str] = None,
+        take: Optional[int] = None,
+        return_count: Optional[bool] = None,
+    ) -> models.PagedProducts:
+        """Reads a list of products.
+
+        Args:
+            continuation_token: The token used to paginate results.
+            take: The number of products to get in this request.
+            return_count: Whether or not to return the total number of products available.
+
+        Returns:
+            A list of products.
+
+        Raises:
+            ApiException: if unable to communicate with the ``/nitestmonitor`` Service
+                or provided an invalid argument.
+        """
+        ...
+
+    @get("products/{id}")
+    def get_product(self, id: str) -> models.Product:
+        """Retrieves a single product by id.
+
+        Args:
+            id (str): Unique ID of a products.
+
+        Returns:
+            The single product matching `id`
+
+        Raises:
+            ApiException: if unable to communicate with the ``/nitestmonitor`` Service
+                or provided an invalid argument.
+        """
+        ...
+
+    @post("query-products")
+    def query_products(
+        self, query: models.QueryProductsRequest
+    ) -> models.PagedProducts:
+        """Queries for products that match the filter.
+
+        Args:
+            query : The query contains a DynamicLINQ query string in addition to other details
+                about how to filter and return the list of products.
+
+        Returns:
+            A paged list of products with a continuation token to get the next page.
+
+        Raises:
+            ApiException: if unable to communicate with the ``/nitestmonitor`` Service or provided invalid
+                arguments.
+        """
+        ...
+
+    @returns.json  # type: ignore
+    @post("query-product-values")
+    def query_product_values(
+        self, query: models.QueryProductValuesRequest
+    ) -> List[str]:
+        """Queries for products that match the query and returns a list of the requested field.
+
+        Args:
+            query : The query for the fields you want.
+
+        Returns:
+            A list of the values of the field you requested.
+
+        Raises:
+            ApiException: if unable to communicate with the ``/nitestmonitor`` Service or provided
+            invalid arguments.
+        """
+        ...
+
+    @post("update-products", args=[Field("products"), Field("replace")])
+    def update_products(
+        self, products: List[Product], replace: bool = False
```

Comment:
```suggestion
        self, products: List[Product], replace: Optional[bool] = False
```

Reply:
I've tried several times to make this work and ran into no end of problems with type casting and subclasses. I think that we don't technically need to mark this as optional. It's not _wrong_ to pass a False from this layer. You're correct that it is _allowed_ not to provide this to the REST layer, but providing False by default is ok. If you can show me how to fix this, we can do it, otherwise, I propose this is ok.

Reply:
https://github.com/prkumar/uplink/issues/278

I filed an issue for this a couple years ago 🙁